### PR TITLE
Bugfix producer pervote share update

### PR DESCRIPTION
**Pull request by me-vadim**
Fixed how pervote shares are updated when votes are changed
Add tests

### CODE REVIEW

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 1]
```diff
@@ -125,7 +137,7 @@ namespace eosiosystem {
             _producers.modify(prod, same_payer, [&](auto& p) {
                p.last_expected_produced_blocks_update = timestamp;
             });
-            _gstate.last_schedule[i] = std::make_pair(prod_name, prod.total_votes / _gstate.total_active_producer_vote_weight);
+            _gstate.last_schedule[i] = std::make_pair(prod_name, (std::floor(prod.total_votes * 10.0) / 10.0) / _gstate.total_active_producer_vote_weight);
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/producer_pay.cpp]: should we call update_pervote_shares() here?

Comment (me-vadim) [reply contracts/contracts/rem.system/src/producer_pay.cpp]: we can, but this will add one more loop over _gstate.last_schedule

**Review by AndriiAndrusyshyn — APPROVED** [review]

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 2]
```diff
@@ -23,24 +25,30 @@ namespace eosiosystem {
...
       for (const auto& p: _gstate.last_schedule) {
-         const auto reward = std::min(int64_t(amount * p.second), amount_remained);
-         amount_remained -= reward;
+         const auto reward = int64_t(amount * p.second);
          total_reward_distributed += reward;
          const auto& prod = _producers.get(p.first.value);
          _producers.modify(prod, eosio::same_payer, [&](auto& p) {
             p.pending_pervote_reward += reward;
          });
-         if (amount_remained == 0) {
```
Comment (AndriiAndrusyshyn) [inline contracts/contracts/rem.system/src/producer_pay.cpp:36]: May be, you can use smth like ``` !amount_remained ```?

File: contracts/contracts/rem.system/src/voting.cpp [hunk 3]
```diff
@@ -248,8 +249,8 @@ namespace eosiosystem {
             const auto active_prod = std::find_if(std::begin(_gstate.last_schedule), std::end(_gstate.last_schedule),
                [target = pd.first](const auto& prod){ return prod.first.value == target.value; });
             if (active_prod != std::end(_gstate.last_schedule)) {
-               _gstate.total_active_producer_vote_weight += pd.second.first;
-               active_prod->second = pitr->total_votes / _gstate.total_active_producer_vote_weight;
+               _gstate.total_active_producer_vote_weight += pitr->total_votes - total_votes_before;
+               update_pervote_shares();
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/voting.cpp]: @me-vadim can we move this out of the loop?

**Review by AndriiAndrusyshyn — APPROVED** [review]

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 4]
```diff
@@ -125,7 +133,8 @@ namespace eosiosystem {
             _producers.modify(prod, same_payer, [&](auto& p) {
                p.last_expected_produced_blocks_update = timestamp;
             });
-            _gstate.last_schedule[i] = std::make_pair(prod_name, prod.total_votes / _gstate.total_active_producer_vote_weight);
+            const double share = prod.total_votes / _gstate.total_active_producer_vote_weight;
+            _gstate.last_schedule[i] = std::make_pair(prod_name, std::floor(share * 100000.0) / 100000.0);
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/producer_pay.cpp]: @me-vadim could you please leave a comment why do we need such rounding

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 5]
```diff
@@ -23,24 +25,30 @@ namespace eosiosystem {
...
-            break;
-         }
       }
       check(total_reward_distributed <= amount, "distributed reward above the given amount");
       return total_reward_distributed;
    }
 
+   void system_contract::update_pervote_shares()
+   {
+      for (auto& p: _gstate.last_schedule) {
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/producer_pay.cpp]: @me-vadim please use `std::for_each`
```
std::for_each( _gstate.last_schedule.bing(), _gstate.last_schedule.end()
                []{ auto p& ) {
                        const auto& prod_name = p.first;
                        const auto& prod = _producers.get(prod_name.value);
                        const double share = prod.total_votes / _gstate.total_active_producer_vote_weight;
                        p.second = std::floor(share * 100000.0) / 100000.0;
                }
);
```

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 6]
```diff
@@ -284,7 +293,7 @@ namespace eosiosystem {
       const auto to_per_vote_pay  = share_pervote_reward_between_producers(amount.amount * 0.2); //TODO: move to constants section
       const auto to_rem           = amount.amount - (to_per_stake_pay + to_per_vote_pay);
       if( amount.amount > 0 ) {
-        token::transfer_action transfer_act{ token_account, { {_self, active_permission} } };
+        token::transfer_action transfer_act{ token_account, { {payer, active_permission} } };
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/producer_pay.cpp:295]: @me-vadim `_self` means contract owner here?

Comment (me-vadim) [reply contracts/contracts/rem.system/src/producer_pay.cpp:295]: yes

File: contracts/contracts/rem.system/src/producer_pay.cpp (hunk 5 above)
Comment (me-vadim) [reply contracts/contracts/rem.system/src/producer_pay.cpp]: what is the reason?

File: contracts/contracts/rem.system/src/voting.cpp (hunk 3 above)
Comment (me-vadim) [reply contracts/contracts/rem.system/src/voting.cpp]: Done

File: contracts/contracts/rem.system/src/producer_pay.cpp (hunk 4 above)
Comment (me-vadim) [reply contracts/contracts/rem.system/src/producer_pay.cpp]: Done

File: contracts/contracts/rem.system/src/producer_pay.cpp (hunk 5 above)
Comment (me-vadim) [reply contracts/contracts/rem.system/src/producer_pay.cpp]: Done

**Review by mark-tsyrulnyk — APPROVED** [review]

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 7]
```diff
@@ -125,7 +136,9 @@ namespace eosiosystem {
             _producers.modify(prod, same_payer, [&](auto& p) {
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/producer_pay.cpp:133]: what if we move this `modify` to the previous `for` and then replace this `for` with the call to `update_pervote_shares()`?

Comment (me-vadim) [reply contracts/contracts/rem.system/src/producer_pay.cpp:133]: Done

**Review by roman-tik — APPROVED** [review]

**Review by olexiybuyanskyy — APPROVED** [review]

**Pull request merged**